src/_pytest/main.py

```diff
@@ -367,6 +365,22 @@ def pytest_runtestloop(session: Session) -> bool:
     return True
 
 
+def _in_build(path: Path) -> bool:
+    """Attempt to detect if ``path`` is the root of a buildsystem's artifacts
+    by checking known dirnames patterns, and the presence of configuration in
+    the parent dir by checking for a setup.py, setup.cfg, or pyproject.toml.
+    """
+    if not path.is_dir():
+        return False
+
+    if any(fnmatch_ex(pat, path) for pat in ("build", "dist")):
+        indicators = ("setup.py", "setup.cfg", "pyproject.toml")
```

Comment:
It is important of understand that `setup.cfg` can be an indicator only in case `pyproject.toml` with setuptools pointers exists, or `setup.py` is present.

So the check would need to be more involved.

Reply:
I'd like to verify my understanding of the logic required. I've re-written it and the tests to reflect how I currently understood the discussion here and in #12625, basically, we first need to see `setup.cfg`, and then only if `setup.py` or `pyproject.toml` with appropriate setuptools pointers exist then we'll ignore collecting `build` and `dist`